Sources/NIOSingletonResources/SingletonType.swift

```diff
@@ -0,0 +1,75 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(Darwin)
+import Darwin
+#elseif os(Windows)
+import ucrt
+import WinSDK
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#else
+#error("Unsupported C library")
+#endif
+import NIOCore
+
+/// Singleton resources provided by SwiftNIO for programs & libraries that don't need full control over all operating system resources.
+///
+/// SwiftNIO allows and encourages the precise management of all operating system resources such as threads and file descriptors.
+/// Certain resources (such as the main ``EventLoopGroup``) however are usually globally shared across the program. This means
+/// that many programs have to carry around an ``EventLoopGroup`` despite the fact the don't require the ability to fully return
+/// all the operating resources which would imply shutting down the ``EventLoopGroup``. This type is the global handle for singleton
+/// resources that applications (and some libraries) can use to obtain never-shut-down singleton resources.
+///
+/// Programs and libraries that do not use these singletons will not incur extra resource usage, these resources are lazily initialized on
+/// first use.
+public enum NIOSingletons {}
+
+extension NIOSingletons {
+    /// How many threads will be used for the multi-threaded ``EventLoopGroup``s.
+    ///
+    /// The thread count is ``System/coreCount`` unless the environment variable `NIO_SINGLETONS_MULTI_THREADED_ELG_LOOP_COUNT`
+    /// is set.
+    public static var suggestedMultiThreadedEventLoopGroupLoopCount: Int {
+        return globalSuggestedMultiThreadedEventLoopGroupLoopCount
+    }
+
+    /// How many threads will be used for the blocking `NIOThreadPool`s.
+    ///
+    /// The thread count is determined automatically unless the environment variable `NIO_SINGLETONS_BLOCKING_POOL_THREAD_COUNT` is set.
+    public static var suggestedBlockingPoolThreadCount: Int {
+        return globalSuggestedBlockingPoolThreadCount
+    }
```

Comment:
I think if we're doing this, we should make these _settable_. Otherwise these numbers aren't very helpful as public API and we should drop them. But overriding their default behaviours is a reasonable thing to want to do at program startup.

Reply:
Okay, that's quite interesting. I think I like that.

So essentially
- What the user sets (assuming it's "early enough")
- If user hasn't set anything use what's in the `NIO_bla_blah_COUNT` env var
- If the env var isn't set fall back onto `System.coreCount`?

Reply:
Yup, that works for me.

Reply:
Do we want to fallback to `System.coreCount`? I would have expected to fallback to just `1`?

Reply:
@Lukasa Thanks, that's now implemented